src/charm.py

```diff
@@ -245,22 +251,31 @@ def _get_s3_env(self) -> typing.Dict[str, typing.Any]:
 
         return s3_env
 
-    def _create_discourse_environment_settings(self) -> typing.Dict[str, typing.Any]:
-        """Create a layer config based on our current configuration.
+    def _get_redis_relation_data(self) -> typing.Tuple[typing.Any, typing.Any]:
+        """Get the hostname and port from the redis relation data.
 
         Returns:
-            Dictionary with all the environment settings.
+            Tuple with the hostname and port of the related redis
         """
-        # Get redis connection information from the relation.
-        redis_hostname = None
-        redis_port = 6379
         # This is the current recommended way of accessing the relation data.
         for redis_unit in self._stored.redis_relation:  # type: ignore
-            redis_hostname = self._stored.redis_relation[redis_unit].get("hostname")  # type: ignore
-            redis_port = self._stored.redis_relation[redis_unit].get("port")  # type: ignore
+            # mypy fails to see that this is indexable
+            redis_unit_data = self._stored.redis_relation[redis_unit]  # type: ignore
+            redis_hostname = redis_unit_data.get("hostname", "")  # type: ignore
+            redis_port = redis_unit_data.get("port", 6379)  # type: ignore
             logger.debug(
                 "Got redis connection details from relation of %s:%s", redis_hostname, redis_port
             )
+        return (redis_hostname, redis_port)
+
+    def _create_discourse_environment_settings(self) -> typing.Dict[str, typing.Any]:
+        """Create a layer config based on our current configuration.
+
+        Returns:
+            Dictionary with all the environment settings.
+        """
+        database_relation_data = self._database.get_relation_data()
```

Comment:
Is guaranteed that for every _create_discourse_environment_settings call the get_relation_data is not None?

Reply:
When I wrote this initially, I thought that it was covered by the numerous calls to `_are_db_relations_ready()`. That function makes sure using the `_is_relation_ready()` from the database module that the relation is OK.  

But it's actually not the case.  
Here is the calling tree of `_create_discourse_environment_settings()`:
```
_create_discourse_environment_settings
    _config_force_https
        _config_changed [OK]
            __init__
    _create_layer_config
        _reload_configuration [*]
            _config_changed [OK]
                __init__
            _on_database_relation_broken [KO]
                __init__
            _on_database_endpoints_changed [OK]
                __init__
            _on_database_created [OK]
                __init__
            _on_redis_relation_changed [OK]
                __init__
    _on_add_admin_user_action [KO]
        __init__
    _on_anonymize_user_action [KO]
        __init__
    _set_up_discourse [OK]
        __init__
```

When there's `[OK]` it means that there's a `_are_db_relations_ready()` check. Otherwise it's `[KO]`.  
The actions would be easy to fix.  
What comes after `_reload_configuration()` is subject to `_is_setup_completed()`. The only issues is with `_on_database_relation_broken()`.  
When we brake the relation, we want to update the config even if `get_relation_data()` returns `None`.  

So I decided that the correct thing to do is to always return a well formed `dict[str, str]` with `get_relation_data()` and to change `is_relation_ready()` accordingly. The idea is that we should be able to change the config/layer's discourse at *any* time. The charm has to adapt to these changes.

Reply:
For this PR is a nice solution.
I think this might be more clear/revised when working with the Observer pattern so you can even create an exception for this case.